### DISCUSSION TITLE
drain-cleaner not invoked when nodes are drained.

### DISCUSSION THREAD

**Discussion by raj-saxena** [strimzi/strimzi-kafka-operator#10101] [General]
We installed drain-cleaner recently to smoothen the process of upgrading node-pools using the helm-chart. 

We are using GKE and our cluster has `cert-manager` so we went with default configuration and the only values that we had to change were:
```yaml
namespace:
  create: false
  name: strimzi
```

I have verified that there is a ValidatingWebhookConfiguration - 
```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingWebhookConfiguration
metadata:
  annotations:
    cert-manager.io/inject-ca-from: strimzi/strimzi-drain-cleaner
    meta.helm.sh/release-name: strimzi-drain-cleaner
    meta.helm.sh/release-namespace: strimzi
  creationTimestamp: "2023-06-09T06:16:44Z"
  generation: 4
  labels:
    app: strimzi-drain-cleaner
    app.kubernetes.io/instance: strimzi-drain-cleaner
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: strimzi-drain-cleaner
    app.kubernetes.io/version: 0.4.2
    helm.sh/chart: strimzi-drain-cleaner-0.4.2
  name: strimzi-drain-cleaner
  resourceVersion: "829436629"
  uid: efe30ac2-ad9c-4d32-bfff-8f2b10ec1f2a
webhooks:
- admissionReviewVersions:
  - v1
  clientConfig:
    caBundle: LS0t....0tLQo=
    service:
      name: strimzi-drain-cleaner
      namespace: strimzi
      path: /drainer
      port: 443
  failurePolicy: Ignore
  matchPolicy: Equivalent
  name: strimzi-drain-cleaner.strimzi.io
  namespaceSelector: {}
  objectSelector: {}
  rules:
  - apiGroups:
    - ""
    apiVersions:
    - v1
    operations:
    - CREATE
    resources:
    - pods/eviction
    scope: Namespaced
  sideEffects: None
  timeoutSeconds: 5
```

However, we drain the nodes, we don't see any logs [as this blog says](https://strimzi.io/blog/2021/09/24/drain-cleaner/)
![image](https://github.com/strimzi/drain-cleaner/assets/6253803/5a864dfe-cd1d-49da-930d-04b80c8bb0ba)

Any suggestions to investigate?

**Comment by scholzj** [comment]
How do you drain the nodes? With `kubectl drain` as done in the blog posts? Or with some special GKE tooling? In general, the Kubernetes logs should have any logs if there are some issues. The Drain Cleaner itself is just an HTTP server which is receiving the events form Kubernetes. So there might be two issues -> does it emit the eviction events? And if it does, why are they not sent to the webhook.

**Reply by raj-saxena** [reply]
Yes, I used the `kubectl drain` command to train the nodes. 
I checked the GKE audit logs and noticed the following logs - 
<img width="729" alt="Screenshot 2023-06-15 at 18 42 26" src="https://github.com/strimzi/drain-cleaner/assets/6253803/1c7ed70a-0713-4384-a659-c4c65b548b2e">
What's weird is that the drainer didn't report receiving any requests. 

I also see that the log's response part says there were too many requests as the `reason`.
<img width="501" alt="image" src="https://github.com/strimzi/drain-cleaner/assets/6253803/6d2d7d66-c8a0-4803-a27e-417cfca49c31">
Could it be because of some noise due to the eviction of other pods?

I was trying to drain multiple nodes at the same time. So, that activity is also a suspect. 
I'll try reproducing the issue with just one node and update here.

**Reply by scholzj** [reply]
Hmm, weird. But if it failed, the Kubernetes logs should have more details. I think the controller manager might be the right component.

**Comment by dlouvier** [comment]
Hi @scholzj 

Raj & me we work for the same team and I had some time today to look at this.

I believe it is a problem with [GKE Firewall rules](https://cloud.google.com/kubernetes-engine/docs/how-to/private-clusters#add_firewall_rules) between CP <> AdmissionWebhook Listener Service, nothing related with Strimzi.

Endup with this conclusion because I configured another AdmissionWebhook in parallel to drain-cleaner and I also was not able to see any incoming request. Increased the timeout and it also timeout with a higher value, so it should be sort-kind of firewall aorund for what I've seen in other issues.

**Comment by dlouvier** [comment]
TL;DR: Solution for GKE at the end of the message


Hi again @scholzj

The main problem is [here](https://github.com/strimzi/drain-cleaner/blob/main/helm-charts/helm3/strimzi-drain-cleaner/templates/060-Deployment.yaml#LL28C20-L28C20):

```
            - containerPort: 8443
              name: https
```

The drain-cleaner container is using a _non-standard_ port and by default it seems to be blocked. I looked to another Admission Control implementations (Prometheus Operator and Cert Manager) they expose the port `10250` (kubelet).

```
❯ k get svc strimzi-drain-cleaner
NAME                    TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)   AGE
strimzi-drain-cleaner   ClusterIP   10.10.10.100   <none>        443/TCP   10d
❯ k get endpoints cert-manager-webhook
NAME                   ENDPOINTS           AGE
cert-manager-webhook   10.05.10.92:10250   11d
```

```
❯ k get pod cert-manager-webhook-5cd7ff21d9-hdkjl -o yaml
(...)
    - containerPort: 10250
      name: https
      protocol: TCP
(...)
```

---

- I know from other strimzi components, ports are usually with an `8`. 
- It seems it is not possible to configure the port for `drain-cleaner`

I think it could be worth for other users as well if drain-cleaner will run on port `10250` or at least if this could be configurable, so it is not necessary to add additional firewall rules to default GKE cluster (I bet could happen the same in other clouds)

Best regards,

---

**Solution for GKE Clusters**
([docs](https://cloud.google.com/kubernetes-engine/docs/how-to/private-clusters#add_firewall_rules))

```
❯ gcloud compute firewall-rules create RULE_NAME \
    --action ALLOW \
    --direction INGRESS \
    --source-ranges CIDR_CONTROL_PLANE_MASTER \
    --rules tcp:8443 \
    --target-tags DST_TAGS
```

**Reply by scholzj** [reply]
TBH, I think that port 443 is the standard one for HTTPS communication 🤷 - not 10250 which is just a Kubelet port. And I don't think we can change the default as if others follow a similar logic, it might break it for someone else. That said, I think all that matters is the service port here which is where the control plane should connect (not 8443 in the container). So I think all you should need is to simply change the service and the webhook configuration to any port you want.

**Reply by raj-saxena** [reply]
> I think all that matters is the service port here which is where the control plane should connect (not 8443 in the container)

Yeah, this was our assumption too but we had to upgrade the cluster again recently and I can confirm that we needed to add a firewall rule for allowing traffic from GKE master range to the nodes for the port `8443`. 
We believe that this is because the `Service` is a collection of network `Endpoints` that direct traffic to port `8443`.
![image](https://github.com/strimzi/drain-cleaner/assets/6253803/94d16eb5-db5e-488f-babf-1233329035b6)
(These rules are needed for any communication. So, we have a rule for `cert-manager` too)

**Comment by mike-serchenia** [comment]
We have the same issue, the problem is specifically with node pool upgrade and not `kubectl drain`

during node pool upgrade gke doesn't  send `"operation":"CREATE","kind":"policy/v1, Kind=Eviction"` if kafka spec has maxUnavailable eq 0 (PDB minAvailable: equal to number of kafka replicas)
